### PR TITLE
Add custom global allocator

### LINKED ISSUE #27: Use custom allocator in Python build
When compiling in native Rust, we have the option to include mimalloc or smalloc for improved performance.
In my experience mimalloc and smalloc behave quite similar.

Polars uses mimalloc in the Python binary:
https://github.com/pola-rs/polars/blob/master/py-polars/Cargo.toml#L22

### PR DESCRIPTION
**Pull request by matthewmturner**
Closes #27

### CODE REVIEW

**Comment by Dandandan** [comment]
Did you also try mimalloc  (see `https://github.com/pola-rs/polars/blob/master/py-polars/Cargo.toml#L22`). Should give similar results, might be faster or slower on some queries. Still interested which one is the best to pick.

File: Cargo.toml [hunk 1]
```diff
@@ -35,6 +35,7 @@ datafusion = { version = "^7.0.0", features = ["pyarrow"] }
 datafusion-expr = { version = "^7.0.0" }
 datafusion-common = { version = "^7.0.0", features = ["pyarrow"] }
 uuid = { version = "0.8", features = ["v4"] }
+snmalloc-rs = {version = "0.2", features = ["cache-friendly"]}
```
Comment (houqp) [inline Cargo.toml:38]: the `cache-friendly` feature has been dropped from upstream in newer releases I believe.

**Comment by matthewmturner** [comment]
mimalloc results below - faster on almost all queries except Q8 and Q10.  This may be the better option.
```
q1: 0.0366870829999999
q2: 0.32823987499999996
q3: 1.1415230410000001
q4: 0.028246750000000098
q5: 1.1575570830000004
q6: 1.2152348750000006
q7: 1.1023409590000002
q8: 2.9869192910000004
q9: 0.5675542499999988
q10: 25.437209041
```

**Review by Dandandan — APPROVED** [review]

**Comment by matthewmturner** [comment]
@Dandandan to confirm - I haven't updated PR with mimalloc yet. Did you approve for snmalloc or mimalloc?

**Comment by Dandandan** [comment]
Ah I didn't catch that 😅. I think mimalloc would be a good choice to go forward with. I think it will be interesting in the future to keep experimenting with this (e.g. there is an upcoming smalloc 2 version).

**Comment by matthewmturner** [comment]
Ok sounds good!  Will update.

**Review by Jimexist — APPROVED** [review]

**Pull request merged**